src/main/java/org/eclipsefoundation/marketplace/service/impl/DefaultPromotionService.java

```diff
@@ -0,0 +1,171 @@
+	int promoCount;
+	@ConfigProperty(name = MicroprofilePropertyNames.PROMO_WEIGHT_DEFAULT, defaultValue = "1")
+	int defaultWeight;
+
+	@Inject
+	MongoDao dao;
+
+	@Inject
+	DtoFilter<Listing> listingFilter;
+	@Inject
+	CachingService<List<Listing>> listingCache;
+
+	@Inject
+	DtoFilter<Promotion> promotionFilter;
+	@Inject
+	CachingService<List<Promotion>> promoCache;
+
+	// random used for shuffling collections
+	private Random r = new Random();
+
+	@Override
+	public List<Listing> getListingsForPromotions(RequestWrapper wrapper, List<Promotion> promos) {
+		if (promos == null || promos.isEmpty()) {
+			LOGGER.debug("No promotions were passed, returning empty list");
+			return Collections.emptyList();
+		}
+		// create mapping to get a list of specific IDs, and to add context to the call
+		// for caching
+		Map<String, List<String>> adds = new HashMap<>();
+		adds.put("type", Arrays.asList("Listing"));
+		adds.put(UrlParameterNames.IDS.getParameterName(),
+				promos.stream().map(Promotion::getListingId).collect(Collectors.toList()));
+
+		MongoQuery<Listing> q = new MongoQuery<>(null, adds, listingFilter);
+		// retrieve the possible cached object
+		Optional<List<Listing>> cachedResults = listingCache.get("promo|listings", wrapper, adds,
+				() -> StreamHelper.awaitCompletionStage(dao.get(q)));
+		if (!cachedResults.isPresent()) {
+			LOGGER.error("Error while retrieving cached promotion listings");
+			return Collections.emptyList();
+		}
+
+		// return the results as a response
+		return cachedResults.get();
+	}
+
+	@Override
+	public void injectPromotions(RequestWrapper wrapper, List<Listing> listings) {
+		// create an empty promo query to get all promos
+		MongoQuery<Promotion> q = new MongoQuery<>(null, Collections.emptyMap(), promotionFilter);
+		// retrieve the possible cached object
+		Optional<List<Promotion>> cachedResults = promoCache.get("all|promo", wrapper, Collections.emptyMap(),
+				() -> StreamHelper.awaitCompletionStage(dao.get(q)));
+		if (!cachedResults.isPresent() || cachedResults.get().isEmpty()) {
+			LOGGER.debug("Could not find any promotions to inject, returning");
+			return;
+		}
+		// make a copy of the array to not impact cached values
+		List<Promotion> promos = new ArrayList<>(cachedResults.get());
+		List<Promotion> promoHolding = new ArrayList<>(promoCount);
+		LOGGER.debug("Found {} promotions, maximum number to inject {}", promos.size(), promoCount);
+		// check each promotion to see if it should be injected
+		Promotion curr;
+		while ((curr = getWeightedPromotion(promos)) != null && promoHolding.size() <= promoCount) {
+			// create a local final field referencing the current promotion for stream ref
+			final Promotion p = curr;
+			LOGGER.debug("Checking promo {}", p.getListingId());
+			// check if current promo matches any of the listing IDs
+			if (promos.stream().noneMatch(l -> l.getId().equals(p.getListingId()))) {
```

Comment:
So, `curr` is added to `promoHolding` only if there is no other promotion left in the `promos` list that have the same listingId? Is that correct? If yes, why?

Reply:
The current check will compare the current promotions listing ID with the listings that we were passed. It will only add the promo to the holding list if the listing it represents isn't already in the list. This way we won't get promotions for something we'd already see. We can remove this logic, but it means we might see the same listing twice